dashboard/test/models/concerns/user/level_progressable_test.rb

```diff
@@ -0,0 +1,671 @@
+    let(:level2) {script.script_levels.second.level}
+    let(:level3) {script.script_levels.third.level}
+
+    context 'when no levels have progress' do
+      it 'returns the number of progression levels' do
+        _num_unpassed_progression_levels.must_equal 3
+      end
+    end
+
+    context 'when some levels are unpassed' do
+      before do
+        UserLevel.create(
+          user: user,
+          level: level1,
+          script: script,
+          attempts: 1,
+          best_result: Activity::MINIMUM_PASS_RESULT
+        )
+        UserLevel.create(
+          user: user,
+          level: level2,
+          script: script,
+          attempts: 1,
+          best_result: Activity::MINIMUM_FINISHED_RESULT
+        )
+      end
+
+      it 'returns the number of unpassed progression levels' do
+        # The levels are:
+        # level1 - passed
+        # level2 - failed (unpassed)
+        # level3 - no progress (unpassed)
+        _num_unpassed_progression_levels.must_equal 2
+      end
+    end
+
+    context 'when all levels are passed' do
+      before do
+        # Set all levels as passed
+        [level1, level2, level3].each do |level|
+          UserLevel.create(
+            user: user,
+            level: level,
+            script: script,
+            attempts: 1,
+            best_result: Activity::MINIMUM_PASS_RESULT
+          )
+        end
+      end
+
+      it 'returns 0 unpassed progression levels' do
+        _num_unpassed_progression_levels.must_equal 0
+      end
+    end
+  end
+
+  describe '#unpassed_progression_level?' do
+    subject(:unpassed_progression_level?) {user.unpassed_progression_level?(script_level, user_levels)}
+
+    let(:user) {create :user}
+    let(:script) {create(:script, :with_levels, levels_count: 3)}
+    let(:script_level) {script.script_levels.first}
+
+    let(:level) {script_level.level}
+    let(:user_level_1) {create(:user_level, user: user, level: level, attempts: 1, best_result: Activity::MINIMUM_PASS_RESULT)}
+    let(:user_level_2) {create(:user_level, user: user, level: level, attempts: 1, best_result: Activity::MINIMUM_FINISHED_RESULT)}
+    let(:user_levels) {[user_level_1, user_level_2]}
+
+    context 'when script level is valid progression level' do
+      it 'returns false when at least one user level is passing' do
+        _unpassed_progression_level?.must_equal false
+      end
+
+      it 'returns true when none of the user levels are passing' do
+        user_level_1.update(best_result: Activity::MINIMUM_FINISHED_RESULT)
+        _unpassed_progression_level?.must_equal true
+      end
+    end
+
+    context 'when script level is not a valid progression level' do
+      before do
+        allow(script_level).to receive(:valid_progression_level?).and_return(false)
+      end
+
+      it 'returns false' do
+        _unpassed_progression_level?.must_equal false
+      end
+    end
+  end
+
+  describe '#script_level_hidden?' do
```

Comment:
I'm still seeing some `script_level_hidden?` tests in `user_test.rb`; were those supposed to be removed?

Reply:
I may have missed it. Thanks for checking!!